src/documents/models.py

```diff
@@ -314,6 +397,83 @@ def thumbnail_file(self):
     def thumbnail_url(self):
         return reverse("fetch", kwargs={"kind": "thumb", "pk": self.pk})
 
+    def set_filename(self, filename):
+        if os.path.isfile(self.filename_to_path(filename)):
+            self.filename = filename
+
+
+def delete_empty_directory(directory):
+    if len(os.listdir(directory)) == 0:
+        try:
+            os.rmdir(directory)
+        except os.error:
+            # Directory not empty
+            pass
+
+
+@receiver(models.signals.m2m_changed, sender=Document.tags.through)
+@receiver(models.signals.post_save, sender=Document)
+def update_filename(sender, instance, **kwargs):
+    if instance.filename is None:
+        return
+
+    # Build the new filename
+    new_filename = instance.source_filename_new()
+
+    # If the filename is the same, then nothing needs to be done
+    if instance.filename is None or \
```

Comment:
The `instance.filename is None` check is duplicated, you already checked it above.

Reply:
done